Add endpoint plugin

upcreid opened the pull request:
set an endpoint to get a defcon level.
The endpoint should have the info {...'defcon':[1-5],...}
Another endPoint can be setup to force the level

upcreid committed:
Add endpoint plugin

set an endpoint to get a defcon level.
The endpoint should have the info {...'defcon':[1-5],...}
Another endPoint can be setup to force the level

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+class EndpointPlugin(base.Plugin):
+    """DefCon Endpoint plugin.
+
+    set an endpoint to get a defcon level.
```
Set

upcreid replied:
done

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+            config = {}
+
+        self.url = config.get('url')
+        self.force_url = config.get('force_url', 'http://defcon.com/status')
```
this looks like a weird default, None would probably be better.

upcreid replied:
weird default => None, done

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+    {
+        'title': 'Defcon Production',
+        'url': 'http://defcon.com/api/simple/production/', // Url to endpoint
+        'link_status': 'http://defcon.com/status/production/ // Url to endpoint info
```
Maybe just add `url` and the rest can be guessed from the response

upcreid replied:
done, only url

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+        'url': 'http://defcon.com/api/simple/production/', // Url to endpoint
+        'link_status': 'http://defcon.com/status/production/ // Url to endpoint info
+        'force_url': 'http://defcon.com/api/simple/force/, // Url to force endpoint
+    }
```
explain here that the json returned by the URL must look like the defcon simple API, example:
```
{
    "name": "Escalation",
    "contact": "escalation-team@",
    "link": "https://confluence/ES/Escalation+Home",
    "defcon": 4,
    "description": "fooo"
}
```

upcreid replied:
added

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+        """Return a status or None."""
+        logging.debug('Handling %s' % (url))
+
+        status = base.Status(
```
if you use the simple API:
title = name
link = link
defcon = defcon
description = description

upcreid replied:
done

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+
+        defcon = 0
+
+        if self.force_url is not None:
```
I do not really understand what force_url is trying to do. Can't you get away with only self.url ? (make sure you can configure the defcon level when this is failing).

upcreid replied:
Since settings.py for each component is read only at the init, we need to be able to force a state.
This is what force_url stand for.
What we need is listend to a bunch of endpoint and determine the level of each, but if the force flag is here, put the same level everywhere

upcreid replied:
This has been abandoned, only one url, no force allowed.

iksaif reviewed (changes requested):
:+1: Good idea, will allow components to be chained together (if it respect the defcon simple API)

upcreid committed:
Add Endpoint Plugin / improvments
- better explanation
- keep only url for parameters
- remove force_url
- default strange url => None
- Use the simple API for status

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -66,36 +68,24 @@ def statuses(self):
+        try:
+            defcon = self._get_defcon_from_url(self.url)
+        except requests.exceptions.RequestException as e:
+            print(e)
```
use `logging.exception()`

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -66,36 +68,24 @@ def statuses(self):
         status = base.Status(
-            title=self.title,
-            link=self.link_status,
+            title=self.name,
```
don't you want to use the data returned by the endpoint for name/link/description instead ?

upcreid replied:
hum maybe there is something i 'm missing here
But with the @property, i need to add a setter to modify the name/link/description ?
I'm not able to assign it

iksaif replied:
Why do you want o modify the name link description in` _to_status` ? (the changes you already did look good)

upcreid replied:
ok get it ! sorry ! will change that

upcreid committed:
Add Endpoint Plugin / improvements

- _get_api_from_url get full request response (dict)

upcreid committed:
Add Endpoint Plugin / improvements

- remove blank

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,100 @@
+            reqEndPoint = self._get_api_from_url(self.url)
+
+        except requests.exceptions.RequestException as e:
+            print(e)
```
logging.exception()

upcreid replied:
done

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,100 @@
+
+        except requests.exceptions.RequestException as e:
+            print(e)
+            reqEndPoint = {
```
You probably want to set a better status message in case of failure here (because this is what people will actually see)

upcreid replied:
add "Failed to get info from {}".format(self.url) for all fields

iksaif commented on defcon/plugins/test_endpoint.py:
```diff
@@ -0,0 +1,30 @@
+from defcon.plugins import endpoint
+
+
+class AlertmanagerPluginTests(test.TestCase):
```
EndPointPluginTests

upcreid replied:
done

iksaif commented on defcon/plugins/test_endpoint.py:
```diff
@@ -0,0 +1,30 @@
+    def test_all(self):
+        """Test with some settings."""
+        for api_url in self._API_URLS:
+            p = endpoint.EndpointPlugin(
```
Can you please mock foo.endpoint.io (see examples in other tests)
* once you make it fail (or just use a non-existing hostname, still mock it to avoid network interactions)
* once you make it work

upcreid replied:
two tests
- test_ok: check if all the mandatory fields are here
- test_no_infos: check if the response have all the mandatory, if not send a default payload

i don't have the way to test the 4xx of an url, it's done directly in the endpoint.py

let me know if it's ok

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,100 @@
+                "description": "fooo"
+            }
+
+        status = self._to_status(reqEndPoint, self.url)
```
Rename `reqEndPoint` to `payload` or `data`

upcreid replied:
=> payload

upcreid committed:
Add endpoint plugin / improvements

upcreid reviewed (commented):
changes have been made, let me know !

upcreid committed:
Add Endpoint Plugin / improvments

- corrections

iksaif commented on defcon/plugins/endpoint.py:
```diff
@@ -0,0 +1,102 @@
+            # catch requests exceptions and create a false payload
+            logging.exception(e)
+            payload = {
+                "name": "Failed to get info from {}".format(self.url),
```
That's ok for now, but as a second PR it would be nice to replace that by an exception and to add exception handling in the plugin runner, this way all failing plugins could trigger a status (which would be configurable from settings and have access to the PluginInstance object)

upcreid replied:
Hey, great news !
Ok for what you said in a next PR.
Also i will update the README of the plugins directory
